### DISCUSSION TITLE
How to use use add_slider_widget() with BackgroundPlotter()?

### DISCUSSION THREAD

**Discussion by lamannamarco** [pyvista/pyvista#4262] [Q&A] [answered]
I am currently reading how to use `add_slider_widget()`. The snippet available in PyVista's API Reference on [add_slider_widget()](https://dev.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_slider_widget.html) obviously runs fine, ie I can click on the slider and move it around to increase/decrease the sphere's resolution. 

When replacing `p=pv.Plotter()` with `p=BackgroundPlotter()` (see the MWE below), the slider becomes "static". Dragging it left or right doesn't change the resolution. However, the sphere rotates. Should I be using `add_callback()`? What puzzles me is that one of the inputs corresponds to the time between calls, suggesting that the update would be "automatic", whereas I'd like to be the one moving the slider.

What am I missing? Any suggestion would be greatly appreciated!

If it helps, I am on Mac (OSX 12.6) and I use VS Code (v1.77.1)
 
**MWE**
```
import pyvista as pv
from pyvistaqt import BackgroundPlotter

p = BackgroundPlotter()

def create_mesh(value):
    res = int(value)
    sphere = pv.Sphere(phi_resolution=res, theta_resolution=res)
    p.add_mesh(sphere, name="sphere", show_edges=True)
    return 

slider = p.add_slider_widget(create_mesh,[5, 100],title="Resolution")

p.show()
p.app.exec_()
```

**Comment by banesullivan** [comment]
I'm unable to reproduce an issue with this code. Would you please share a screen capture and a pyvista report:

```py
import pyvista as pv
print(pv.Report())
```


https://user-images.githubusercontent.com/22067021/231484947-f72115b9-f01d-43f6-9560-7aeb91ce46b1.mov

**Comment by lamannamarco** [comment]
Here is the report and the screen capture. 

```
Date: Wed Apr 12 10:26:33 2023 EDT

                OS : Darwin
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 8.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1
       GPU Version : 4.1 Metal - 76.3

  Python 3.8.15 (default, Nov 10 2022, 13:17:42)  [Clang 14.0.6 ]

           pyvista : 0.37.0
               vtk : 9.0.3
             numpy : 1.23.4
           imageio : 2.19.3
            scooby : 0.7.0
             pooch : v1.4.0
        matplotlib : 3.5.3
         pyvistaqt : 0.9.1
             PyQt5 : 5.15.7
           IPython : 8.6.0
             scipy : 1.9.3
              tqdm : 4.64.1

  Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904
  for Intel(R) 64 architecture applications
  ```
  
  https://user-images.githubusercontent.com/49376024/231490779-76f46821-130b-4929-9bfe-f43003ab85e6.mov

**Reply by banesullivan** [reply]
Ah. Looks like the slider is not interactive at all. Can you try upgrading VTK? 9.0 is pretty dated at this point.

**Reply by banesullivan** [reply]
This actually looks like an issue I recall where the interactor messes up the viewport coordinates of the mouse and the event is like clicking somewhere else in the window. I'm pretty sure updating VTK will fix it.

**Comment by lamannamarco** [comment]
First of all, thanks for your replies! The slider is still not interactive, even though I upgraded VTK to its 9.2.6 version (that's the latest version I could find on `conda-forge`). For reference, I have provided the second part of the PyVista report below. 

Do you have any other suggestions I can try?

```
Python 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:05:36)
  [Clang 14.0.6 ]

           pyvista : 0.38.5
               vtk : 9.2.6
             numpy : 1.24.2
           imageio : 2.27.0
            scooby : 0.7.1
             pooch : v1.7.0
        matplotlib : 3.7.1
         pyvistaqt : 0.0.0
             PyQt5 : 5.15.7
           IPython : 8.12.0
        ipywidgets : 8.0.6
             scipy : 1.10.1
              tqdm : 4.65.0
      nest_asyncio : 1.5.6
```
I just wanted to note that it says `pyvistaqt : 0.0.0`. However, if I open terminal, activate my `venv` and do `conda list` I see that the package version is `0.10.0`.

**Reply by banesullivan** [reply]
@pyvista/pyvistaqt-admin, have you all seen/resolved this issue before?

**Reply by larsoner** [reply]
See above about resolution: https://github.com/pyvista/pyvista/discussions/4262#discussioncomment-5593539 . If you put your mouse halfway between where it is and the lower-left corner and do some clicking, does it sometimes show up in the slider? If so it's a HiDPI bug. I have seen these before and would have to look how we solved them (we might have just disabled HiDPI in 3D somehow, which would be a bummer but would at least make things work!)

**Reply by lamannamarco** [reply]
Sorry for the late reply, guys. I have tried clicking in other parts of the window but the slider is still static/inactive. Other things I tried: 

1. Updated my MBA to OSX 13.3.1 -- did not make a difference
2. Created a new `venv` from scratch -- did not make a difference -- in case it's of interest, the `PyVista` report is below

```
Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]

           pyvista : 0.38.5
               vtk : 9.2.6
             numpy : 1.24.2
           imageio : 2.27.0
            scooby : 0.7.1
             pooch : v1.7.0
        matplotlib : 3.7.1
         pyvistaqt : 0.10.0
             PyQt5 : 5.15.9
           IPython : 8.12.0
             scipy : 1.10.1
              tqdm : 4.65.0
      nest_asyncio : 1.5.6
```

**Comment by lamannamarco** [comment]
I am going to close this question. I have not made any changes to the script, but it now works -- which is great! I don't know what solved the issue. When I asked this question, the operating system was `OS X 12.4` and I was working with `VS Code v1.77.1`, `Pyvista 0.38.5`, `vtk 9.2.6` and `pyvistaqt 0.10.0`. Now I have `OS X 13.4`, `VS Code v1.79.2` and `Pyvista 0.39.1`. One of these updates must have helped, but I am unable to pinpoint which one.

**Reply by lamannamarco** [reply, marked as answer]
I don't think the updates solved the problem, because -this week- the slider wasn't working again. However, downgrading PyQt5 from 5.15.9 to 5.13 actually solved the problem. I tried this solution after looking at the following thread: https://gitlab.kitware.com/vtk/vtk/-/issues/17953
